### ISSUE TITLE
make joblib.Parallel return a generator

### ISSUE THREAD

**Issue by drhyrum** [joblib/joblib#217]
Often one wants to perform simple operations on the output of a very long sequence of tasks.  If the number of outputs is large, it may be inefficient or impossible to store them in a list.  Instead, add functionality to joblib.Parallel so that one can do:

``` python
parallel_job = ( delayed( job )( param ) for param in so_many_job_params )  # generator for input
for output in Parallel(n_jobs=10, iterable=parallel_job):                  # generator as output 
   do_something( output )
```

In the example above, I've added the job iterable to the constructor of Parallel.  The only required change would be to add an `__iter__(self)` method to Parallel which has almost identical functionality to `__call__(self.iterable)`, but instead uses `self.iterable` and `yield`s an element one completed job at a time, rather than `return`ing a list of outputs.

**Comment by GaelVaroquaux** [comment]
I would love this. But I believe that last time I looked, I found that it
was actually fairly technical.

**Comment by garbonzuke** [comment]
The following hack seems to work for me, although I admit to not understanding the thread unsafeness of generators...so it may not be as simple as this:
```python
   # modified __init__ to take in an iterable
    def __init__(self, n_jobs=1, backend=None, verbose=0, pre_dispatch='all',
                 temp_folder=None, max_nbytes=100e6, mmap_mode='r', iterable=None):
        self.iterable = iterable
        # ...

   # slightly modified to return a single instance instead of a list
    def retrieve(self):
        output = None
        if self._iterating or len(self._jobs) > 0:  # changed while to if
            if len(self._jobs) == 0:
                # Wait for an async callback to dispatch new jobs
                while self._iterating and len(self._jobs)==0:
                    time.sleep(0.01)
                if not self._iterating:
                    return output
            # We need to be careful: the job queue can be filling up as
            # we empty it
            if hasattr(self, '_lock'):
                self._lock.acquire()
            job = self._jobs.pop(0)
            if hasattr(self, '_lock'):
                self._lock.release()
            try:
                output = job.get()           # <-------  fundamental change here
                return output
            except tuple(self.exceptions) as exception:
                # ....

# created __iter__(self) that is identical to __call__(self,iterable), except for a few lines
    def __iter__(self):
        if self.iterable is None:
            return # nothing to iterate over
        iterable = self.iterable
        
        # like __call__ but returns a generator instead of a list
        # so that all parallel processing need not complete before operating on the output

        # ...

        # inside try:
            # ...
            while self._iterating or len(self._jobs) > 0:     # <-------  fundamental change here
                yield self.retrieve() 
            
            # Make sure that we get a last message telling us we are done
            elapsed_time = time.time() - self._start_time
            
            # commented out the following
            #self._print('Done %3i out of %3i | elapsed: %s finished',
            #            (len(self._output),
            #             len(self._output),
            #                short_format_time(elapsed_time)
            #            ))

            # ...

```

**Comment by aarchiba** [comment]
I don't know whether a generator is the best way to accomplish this, but it would sometimes be very useful to be able to work with the results from whichever tasks happen to have finished. So even a public API that let me get at the list of async_apply objects would be handy.

**Comment by fcharras** [comment]
> I would love this. But I believe that last time I looked, I found that it
> was actually fairly technical.

What would be the issues with @drhyrum code here ?

**Comment by lesteve** [comment]
I had a naive attempt at some point that failed quite miserably. It was hanging if you did not consume fully the iterator and I was not able to figure out why. Look at https://github.com/joblib/joblib/compare/master...lesteve:parallel-return-generator for more details. Help in rebasing this PR and make it work would be more than welcome.

**Comment by fcharras** [comment]
Thank you I'd very much like to take a look at it, will update if any improvements. Currently I have in mind to try to feed Parallel a generator that read directly into Parallel._output in order to yield new points.

**Comment by cgarciae** [comment]
I coded this in a library called `pypeln`, however it uses multiprocessing and some libraries on OSX crash because of this so I have to fallback to joblib for that OS.

**Comment by louisabraham** [comment]
I think that using sharedmemory with a queue would provide a solution with little performance cost.

**Comment by leockl** [comment]
Has there been any update on this? Would be really great if we could make joblib.Parallel return a generator rather than a list

**Comment by beasteers** [comment]
I would love this functionality to be added - if not just so you could wrap `Parallel` with a progress bar. I see there are a bunch of closed PRs trying to implement this, but none of them have been merged for some reason :/

In the meantime, if anyone wants a quick drop-in (albeit hacky) solution, this has been working for me. It works without having to copy and make deep edits to the original code.
```python
import joblib
from joblib import delayed
import threading

class Parallel(joblib.Parallel):
    def it(self, iterable):
        try:
            t = threading.Thread(target=self.__call__, args=(iterable,))
            t.start()

            i = 0
            output = self._output
            while t.is_alive() or (output and i < len(output)):
                # catch the list reference and store before it's overwritten
                if output is None:
                    output = self._output
                # yield when a new item appears
                if output and i < len(output):
                    yield output[i]
                    i += 1
        finally:
            t.join()


########
# usage
########

import time
import tqdm

def task(x):
    time.sleep(1)
    return x * 2

for x in Parallel(3).it(delayed(task)(x) for x in range(12)): 
    print(x)
```

**Comment by dave31415** [comment]
Can just mini-batch it with multiple calls to Parallel

**Comment by tomMoral** [comment]
Indeed, this has been implemented in #1393 , closing this long lasting issue :)
